### ISSUE TITLE
`textDocument` has an illegal filesystem URI

### ISSUE THREAD

**Issue by ckipp01** [com-lihaoyi/mill#2393]
When using Metals with a codebase (discovered when working on the scala-cli codebase), you can get a `publishDiagnostics` message that contains an invalid filesystem URI since it contains `?` making it think it's a query. For example you'll see this in the logs:

```
Mar 27, 2023 12:58:28 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification
WARNING: Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "build/publishDiagnostics",
  "params": {
    "textDocument": {
      "uri": "file:///Users/ckipp/Documents/scala-workspace/scala-cli/cli/test?id\u003dcli.test"
    },
    "buildTarget": {
      "uri": "file:///Users/ckipp/Documents/scala-workspace/scala-cli/cli/test?id\u003dcli.test"
    },
    "diagnostics": [
      {
        "range": {
          "start": {
            "line": 0,
            "character": 0
          },
          "end": {
            "line": 0,
            "character": 0
          }
        },
        "severity": 2,
        "code": "-1",
        "source": "mill",
        "message": "Flag -Xsemanticdb set repeatedly"
      },
      {
        "range": {
          "start": {
            "line": 0,
            "character": 0
          },
          "end": {
            "line": 0,
            "character": 0
          }
        },
        "severity": 2,
        "code": "-1",
        "source": "mill",
        "message": "Flag -Xsemanticdb set repeatedly"
      }
    ],
    "reset": true
  }
}
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
        at org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer.consume(TracingMessageConsumer.java:114)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
        ... 12 more
Caused by: java.lang.IllegalArgumentException: URI has a query component
        at java.base/sun.nio.fs.UnixUriUtils.fromUri(UnixUriUtils.java:58)
        at java.base/sun.nio.fs.UnixFileSystemProvider.getPath(UnixFileSystemProvider.java:102)
        at java.base/java.nio.file.Path.of(Path.java:203)
        at java.base/java.nio.file.Paths.get(Paths.java:98)
        at scala.meta.internal.mtags.MtagsEnrichments$XtensionURIMtags.toAbsolutePath(MtagsEnrichments.scala:120)
        at scala.meta.internal.mtags.MtagsEnrichments$XtensionStringMtags.toAbsolutePath(MtagsEnrichments.scala:173)
        at scala.meta.internal.metals.MetalsEnrichments$XtensionString.toAbsolutePath(MetalsEnrichments.scala:595)
        at scala.meta.internal.metals.MetalsEnrichments$XtensionString.toAbsolutePath(MetalsEnrichments.scala:592)
        at scala.meta.internal.metals.Diagnostics.onBuildPublishDiagnostics(Diagnostics.scala:142)
        at scala.meta.internal.metals.clients.language.ForwardingMetalsBuildClient.onBuildPublishDiagnostics(ForwardingMetalsBuildClient.scala:109)
        ... 17 more
```

**To Reproduce**
- Clone scala-cli
- Import the build (using mill-bsp)
- See error in logs

**Expected behaviour**
I wouldn't expect the target URI to be used in the `textDocument.uri` here as this isn't a valid filesystem URI.

**Comment by lefou** [comment]
I once found that bloop is using that kind of URIs and thought, it must be the way to tell a BSP client what the "short" path is. This is probably incorrect. I'm glad to remove that part.

**Comment by ckipp01** [comment]
Hey! Thanks for the quick fix! This looks fine, although I wasn't able to test it in the scala-cli build since there is a bunch of stuff that would need to be published for 0.11.x. However if I see this pop up again or anything funky with it I'll let you know!
